### PR TITLE
Support for creating images

### PR DESCRIPTION
**Pull request by nvnyale**

### CODE REVIEW

File: api/orchestration_images.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	if err != nil {
+		return "", apierror.New(apierror.ErrBadRequest, err.Error(), nil)
+	}
+
+	input := &ec2.CreateImageInput{
+		InstanceId:  req.InstanceId,
+		Name:        req.Name,
+		Description: req.Description,
+	}
+	input.NoReboot = aws.Bool(!aws.BoolValue(req.ForceReboot))
```
Comment (tenyo) [inline api/orchestration_images.go:30]: Why not just put this inside the struct above?

File: api/orchestration_images.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	if aws.BoolValue(req.CopyTags) {
+
```
Comment (tenyo) [inline api/orchestration_images.go:32]: ✂️

File: api/handlers_images.go [hunk 3]
```diff
@@ -144,3 +144,48 @@ func (s *server) ImageUpdateHandler(w http.ResponseWriter, r *http.Request) {
...
 }
+
+func (s *server) ImageCreateHandler(w http.ResponseWriter, r *http.Request) {
+	w = LogWriter{w}
+	vars := mux.Vars(r)
+	account := s.mapAccountNumber(vars["account"])
+
+	req := &Ec2ImageCreateRequest{}
+	if err := json.NewDecoder(r.Body).Decode(req); err != nil {
+		msg := fmt.Sprintf("cannot decode body into update image input: %s", err)
```
Comment (tenyo) [inline api/handlers_images.go:155]: 👀 create

File: api/handlers_images.go [hunk 4]
```diff
@@ -144,3 +144,48 @@ func (s *server) ImageUpdateHandler(w http.ResponseWriter, r *http.Request) {
...
+		handleError(w, apierror.New(apierror.ErrBadRequest, msg, err))
+		return
+	}
+
+	if req.InstanceId == nil || req.Name == nil || req.CopyTags == nil || req.ForceReboot == nil {
```
Comment (tenyo) [inline api/handlers_images.go:160]: Only InstanceId and Name are currently required. We can set defaults for the others if they're not defined.
see https://git.yale.edu/spinup/ec2-api/blob/master/lib/json_schemas/create_ami.json

File: api/handlers_images.go [hunk 5]
```diff
@@ -144,3 +144,48 @@ func (s *server) ImageUpdateHandler(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	imageId, err := orch.createImage(r.Context(), req)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	w.Write([]byte(imageId))
```
Comment (tenyo) [inline api/handlers_images.go:190]:
```
handleResponseOk(w, out)
```

File: api/handlers_images.go [hunk 6]
```diff
@@ -144,3 +144,48 @@ func (s *server) ImageUpdateHandler(w http.ResponseWriter, r *http.Request) {
...
+		policyArns: []string{
+			"arn:aws:iam::aws:policy/AmazonEC2ReadOnlyAccess",
+		},
+	})
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+
+	imageId, err := orch.createImage(r.Context(), req)
```
Comment (tenyo) [inline api/handlers_images.go:183]: Let's call this `out` instead of `imageId`

**Review by tenyo — COMMENTED** [review]
A few small things - after we fix them let's try it in test.

File: api/handlers_images.go (hunk 5 above)
Comment (nvnyale) [reply api/handlers_images.go:190]: I thought about it but `handleResponseOk` doesn't send the status 201. Please let me know your thoughts.

Comment (tenyo) [reply api/handlers_images.go:190]: We don't really need the 201, let's standardize on 200 and use the `handleResponseOk`

File: api/orchestration_images.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/aws/aws-sdk-go/service/ec2"
+	log "github.com/sirupsen/logrus"
+)
+
+func (o *ec2Orchestrator) createImage(ctx context.Context, req *Ec2ImageCreateRequest) (string, error) {
+	if req == nil {
+		return "", apierror.New(apierror.ErrBadRequest, "invalid input", nil)
+	}
+
+	log.Debugf("got request to create instance: %s", awsutil.Prettify(req))
```
[the remark this answers is not in the document]
Comment (nvnyale) [reply api/orchestration_images.go]: Done.

File: api/handlers_images.go (hunk 3 above)
Comment (nvnyale) [reply api/handlers_images.go]: Done.

File: api/handlers_images.go (hunk 6 above)
Comment (nvnyale) [reply api/handlers_images.go]: Done.

File: api/handlers_images.go [hunk 8]
```diff
@@ -152,15 +153,18 @@ func (s *server) ImageCreateHandler(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
 
-	if req.InstanceId == nil || req.Name == nil || req.CopyTags == nil || req.ForceReboot == nil {
-		handleError(w, apierror.New(apierror.ErrBadRequest, "missing required fields: instance_id, name, copy_tags, force_reboot", nil))
+	if req.InstanceId == nil || req.Name == nil {
+		handleError(w, apierror.New(apierror.ErrBadRequest, "missing required fields: instance_id, name", nil))
 		return
 	}
+	if req.ForceReboot == nil {
```
Comment (tenyo) [inline api/handlers_images.go:165]: CopyTags should default to true

File: api/handlers_instances.go [hunk 9]
```diff
@@ -412,10 +412,10 @@ func (s *server) InstanceSendCommandHandler(w http.ResponseWriter, r *http.Reque
...
 	}
 
-	if len(req.Parameters) == 0 {
-		handleError(w, apierror.New(apierror.ErrBadRequest, "Parameters are required", nil))
-		return
-	}
+	// if len(req.Parameters) == 0 {
+	// 	handleError(w, apierror.New(apierror.ErrBadRequest, "Parameters are required", nil))
+	// 	return
+	// }
```
Comment (tenyo) [inline api/handlers_instances.go:418]: Why is this commented out?

**Pull request merged**